### PR TITLE
Breaking change: Throw when resolving on a disposed service provider

### PR DESCRIPTION
**Pull request by maryamariyan**
- Resolving Services after ServiceProvider is Disposed should throw with ObjectDisposedException

@davidfowl, for testing I prepared a test similar to the repro code you provided earlier in [this comment](https://github.com/dotnet/runtime/issues/35986#issuecomment-670302230)

Similar to the use case you were illustrating there, this use case (before this PR change) would also result in a deadlock. Here we try to call Dispose while the singleton lock `ResolvedServices` is taken). As shown in this test now it would throw ObjectDisposedException instead.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 1]
```diff
@@ -264,6 +266,70 @@ public void ScopeDispose_PreventsServiceResolution()
...
+            });
+
+           var t2 = Task.Run(() =>
+            {
+                ThreadId = 2;
+                using var scope2 = sp.CreateScope();
+                scope2.ServiceProvider.GetRequiredService<Thing2>();
+            });
+
+            await Assert.ThrowsAsync<ObjectDisposedException>(async () => {await t1; await t2;});
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:317]: nit: indentation
```suggestion
            await Assert.ThrowsAsync<ObjectDisposedException>(async () => { await t1; await t2; });
```

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@ internal class ServiceProviderEngineScope : IServiceScope, IServiceProvider, IAs
         private List<object> _disposables;
 
         private bool _disposed;
+        private static object s_locker = new object();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:20]: This shouldn't be static, it should be an instance.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 3]
```diff
@@ -266,55 +265,62 @@ public void ScopeDispose_PreventsServiceResolution()
...
-                return new Thing1(sp);
+                // foo ctor disposes ServiceProvider
+                var foo = sp.GetRequiredService<DisposableFoo>();
             });
+        }
 
-            services.AddTransient(sp =>
+        public class Foo : IDisposable
+        {
+            public static int TimesConstructorCalled = 0;
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:298]: Why is this static?

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 4]
```diff
@@ -17,6 +17,7 @@ internal class ServiceProviderEngineScope : IServiceScope, IServiceProvider, IAs
         private List<object> _disposables;
 
         private bool _disposed;
+        private object _locker = new object();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:20]: nit: _lock?

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 5]
```diff
@@ -41,17 +42,24 @@ public object GetService(Type serviceType)
...
             {
                 return service;
             }
 
-            lock (ResolvedServices)
+            lock (_locker)
             {
+                if (_disposed)
+                {
+                    if (_disposables != null)
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:56]: dispose service here before throwing.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 6]
```diff
@@ -41,17 +42,24 @@ public object GetService(Type serviceType)
...
+                    {
+                        // cleanup disposable just in case still has items before throwing
+                        _disposables = null;
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:59]: I'm not sure we would do this. This code shouldn't touch the disposables list.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs (hunk 3 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:298]: Yeah it doesn't need to be static. I'm pushing up a cleanup for this diff.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 7]
```diff
@@ -41,17 +42,27 @@ public object GetService(Type serviceType)
...
             }
 
-            lock (ResolvedServices)
+            lock (_lock)
             {
+                if (_disposed)
+                {
+                    if (service is IAsyncDisposable asyncDisposable)
+                    {
+                        // TODO complete
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:58]: - [ ] TODO complete

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:58]: Oh this is tricky.... since this code path is all synchronous.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:58]: We may have to block here... We should also maybe prefer IDisposable over IAsyncDisposable to make this less of a problem.

**Comment by maryamariyan** [comment]
Thanks @davidfowl. I think the PR is ready for another now.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 8]
```diff
@@ -41,17 +42,27 @@ public object GetService(Type serviceType)
...
             {
+                if (_disposed)
+                {
+                    if (service is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                    }
+                    else
+                    {
+                        ((IAsyncDisposable)service).DisposeAsync().AsTask().Wait();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:62]: `GetAwaiter().GetResult()`

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 9]
```diff
@@ -41,17 +42,27 @@ public object GetService(Type serviceType)
...
             {
+                if (_disposed)
+                {
+                    if (service is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                    }
+                    else
+                    {
+                        ((IAsyncDisposable)service).DisposeAsync().AsTask().GetAwaiter().GetResult();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:62]: The more I think about this and potential downsides, I'm thinking maybe we should fire and forget these...

cc @stephentoub for advice

TL;DR We're trying to take more granular locks in DI so the old code used to wait on what ended up being a global lock per service provider instance. Before disposing during resolve would wait, now we want to remove that global lock from resolve we're changing the behavior to let dispose complete and have any new calls to this method throw an ODE and dispose the thing being created.

Disposing the thing being created means we need to support IAsyncDisposable and IDisposable and this is a synchronous code path. I'm thinking we fire and forget since it's an error case and waiting may cause bad side effects, like a deadlock.

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:62]: The two potential problems with fire-and-forget in a case like this are:
1. The thing being disposed (whatever is calling CaptureDisposable) borrowed some resources that end up still being in use when its Dispose returns, and the caller then rightfully assumes they've regained sole ownership but suffers strange race conditions because they're assumption was wrong.
2. Failures that occur asynchronously end up being masked.

In general (1) is the more impactful issue if it's applicable... is it applicable here?

(2) can be mitigated if there's some logging mechanism that could be employed as part of a continuation.

How common is it for a service to implement IAsyncDisposable but not IDisposable?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:62]: > How common is it for a service to implement IAsyncDisposable but not IDisposable?

I only know of one thing that does this https://source.dot.net/#Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs,59 😄 

I do think it's extremely rare though. Maybe we just need to avoid the deadlock and call it a day

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 10]
```diff
@@ -271,9 +271,9 @@ public void GetService_ThenDisposeOnDifferentThread_ServiceDisposed()
         {
             var services = new ServiceCollection();
             services.AddSingleton<Foo1>();
-            var sp = services.BuildServiceProvider();
+            IServiceProvider sp = services.BuildServiceProvider();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:274]: This can be typed as ServiceProvider and the cast can be removed further down.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 11]
```diff
@@ -17,6 +17,7 @@ internal class ServiceProviderEngineScope : IServiceScope, IServiceProvider, IAs
         private List<object> _disposables;
 
         private bool _disposed;
+        private object _lock = new object();
```
Comment (JamesNK) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:20]:
```suggestion
        private readonly object _lock = new object();
```

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 12]
```diff
@@ -41,17 +42,30 @@ public object GetService(Type serviceType)
...
+                    if (service is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                    }
+                    else
+                    {
+                        // this synchronous code path could cause a deadlock if we were to block on DisposeAsync
+                        // instead, a fire-and-forget call may cause race conditions while disposing borrowed resources.
+                        // since this is a rare error case, we just take the fire-and-forget approach here
+                        Task.Run(() => ((IAsyncDisposable)service).DisposeAsync().AsTask().GetAwaiter().GetResult());
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: cc: @stephentoub

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: I don't understand the comment.

But separate from that, if the goal is fire-and- 
forget, then you don't want or need the .GetAwaiter().GetResult(). And the Task.Run shouldn't be necessary either.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: We're not doing fire and forget, we're trying to avoid a deadlock for sync contexts that are single threaded in this rare case when somebody disposes during construction *and* they only implemented `IAsyncDisposable`.

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: > We're not doing fire and forget,

As written,  it is. Maybe the last paren is misplaced.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: @maryamariyan we should update the comment. We're not doing fire and forget but we're doing sync over async here in the rare case where an object only implements IAsyncDisposable we may end up starving the thread pool.

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:65]: OK.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 13]
```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
...
         }
 
+        [Fact]
+        public void GetService_ThenDisposeOnDifferentThread_ServiceDisposed()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<Foo1>();
+            IServiceProvider sp = services.BuildServiceProvider();
+            var foo = sp.GetRequiredService<Foo1>();
+            Task.Run(() => (sp as IDisposable).Dispose()).Wait();
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:277]: Should this instead be an `async Task` test, and then `await Task.Run(() => ...);`? Even though this is tests - it is a good habit to be in.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 14]
```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
...
+            }
+            public bool IsDisposed { get; private set; }
+
+            public void Dispose()
+            {
+                IsDisposed = true;
+            }
+        }
+
+        private class Foo2 : IDisposable
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:339]: I would name this class appropriately. `private class DisposeServiceProviderInCtor`

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 15]
```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
...
+            }).Wait(TimeSpan.FromSeconds(10));
+
+            Assert.True(success);
+        }
+
+        private class Foo1 : IDisposable
+        {
+            public Foo1()
+            {
+                Thread.Sleep(5000);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:329]: What is this testing?

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 16]
```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
...
+                            // ctor disposes ServiceProvider
+                            var foo = sp.GetRequiredService<FooAsyncDisposable>();
+                        });
+                    },
+                    CancellationToken.None, 
+                    TaskCreationOptions.None,
+                    TaskScheduler.FromCurrentSynchronizationContext()
+                    ).Wait();
+                });
+            }).Wait(TimeSpan.FromSeconds(10));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:320]: Why does this need to be in a separate Task? Why can't it just be

```suggestion
                SingleThreadedSynchronizationContext.Run(() =>
                {
                        // Act
                        Assert.Throws<ObjectDisposedException>(() =>
                        {
                            // ctor disposes ServiceProvider
                            var foo = sp.GetRequiredService<FooAsyncDisposable>();
                        });
                });
            }).Wait(TimeSpan.FromSeconds(10));
```

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 17]
```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
...
+            
+            Assert.True(disposable.Disposed);
+            Assert.True(asyncDisposable.DisposeAsyncCalled);
+            if (includeDelayedAsyncDisposable)
+            {
+                Assert.Equal(1, delayedAsyncDisposableService.DisposeCount);
+            }
+        }
+
+        private class FooAsyncDisposable : IFakeService, IAsyncDisposable
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:380]: I would name this class what it does: `private class DisposeServiceProviderInCtorAsyncDisposable`

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs (hunk 15 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:329]: If meanwhile resolving foo service, service provider gets disposed, the test asserts that we dispose of the service as well.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs (hunk 16 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: Thanks. This needed to get simplified.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs (hunk 15 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:322]: But the way the test is written above, this sleep will happen inline and then the object will be returned on line 276.

Once the object is returned, then you are starting a new thread to dispose the service provider.

So I don't think the sleep is doing anything beneficial.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: Good catch, we need to kick off 2 threads.

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: Removed this test and used tests below to also assert the service (for both IDisposable and IAsyncDisposable) gets disposed:

- `GetAsyncService_DisposeAsyncOnSameThread_ThrowsAndDoesNotHangAndDisposeAsyncGetsCalled`
- `GetService_DisposeOnSameThread_ThrowsAndDoesNotHangAndDisposeGetsCalled`

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: - [ ] Add another test case.

**Review by eerhardt — APPROVED** [review]
This looks good to me. Nice work, @maryamariyan.

**Review by davidfowl — APPROVED** [review]
